### ISSUE TITLE
multi-node distributed training

### ISSUE THREAD

**Issue by zhao1157** [horovod/horovod#2568]
horovod version: 0.20.2

I tried on two nodes and renamed the network interfaces of the two nodes so that they were the same. Before starting the docker container on the primary node, I launched a container on the secondary node with `/usr/sbin/sshd -p 22345; sleep infinity`. On the primary node if I ran ` horovodrun -np 2 -H localhost:1,secondary_node:1 -p 22345`, it hung for a while and exited with a warning 

> /usr/lib/python3.6/multiprocessing/semaphore_tracker.py:143: UserWarning: semaphore_tracker: There appear to be 5 leaked semaphores to clean up at shutdown len(cache))

I also tried to increase `--start-timeout`, but with no avail.

After digging into the horovod source code and manually set `--network-interface=enx00e04c36024f`, and it worked. I wonder what causes the hang without explicitly specifying the network interface. 

Another issue is the processing speed of utilizing two nodes is absurdly slower, 0.1 vs 25 images/second. Any idea on how to fix this performance degradation?

**Comment by chongxiaoc** [comment]
Can you run `ifconfig` on each nodes to see what available nics are there? If there is no `--network-interface` specified, horovod is using `ifconfig` and local ip address to pick up available nics internally.

**Comment by zhao1157** [comment]
@chongxiaoc Here is the output of `ifconfig`

On one node:

> br-4f2ca70ca73a: flags=4099<UP,BROADCAST,MULTICAST>  mtu 1500
        inet 172.18.0.1  netmask 255.255.0.0  broadcast 172.18.255.255
        inet6 fe80::42:4bff:fe22:7cb  prefixlen 64  scopeid 0x20<link>
        ether 02:42:4b:22:07:cb  txqueuelen 0  (Ethernet)
        RX packets 439  bytes 34848 (34.8 KB)
        RX errors 0  dropped 0  overruns 0  frame 0
        TX packets 558  bytes 485416 (485.4 KB)
        TX errors 0  dropped 0 overruns 0  carrier 0  collisions 0

> docker0: flags=4163<UP,BROADCAST,RUNNING,MULTICAST>  mtu 1500
        inet 172.17.0.1  netmask 255.255.0.0  broadcast 172.17.255.255
        inet6 fe80::42:5ff:fed2:a8d9  prefixlen 64  scopeid 0x20<link>
        ether 02:42:05:d2:a8:d9  txqueuelen 0  (Ethernet)
        RX packets 16026521  bytes 1202950371 (1.2 GB)
        RX errors 0  dropped 0  overruns 0  frame 0
        TX packets 20490819  bytes 203232121445 (203.2 GB)
        TX errors 0  dropped 0 overruns 0  carrier 0  collisions 0

> enp96s0f0: flags=4099<UP,BROADCAST,MULTICAST>  mtu 1500
        ether b4:05:5d:51:b9:4e  txqueuelen 1000  (Ethernet)
        RX packets 0  bytes 0 (0.0 B)
        RX errors 0  dropped 0  overruns 0  frame 0
        TX packets 0  bytes 0 (0.0 B)
        TX errors 0  dropped 0 overruns 0  carrier 0  collisions 0

> enx00e04c363afe: flags=4163<UP,BROADCAST,RUNNING,MULTICAST>  mtu 1500
        inet 10.166.15.59  netmask 255.255.255.0  broadcast 10.166.15.255
        inet6 fe80::2e0:4cff:fe36:3afe  prefixlen 64  scopeid 0x20<link>
        ether 00:e0:4c:36:3a:fe  txqueuelen 1000  (Ethernet)
        RX packets 22568629  bytes 17593737754 (17.5 GB)
        RX errors 0  dropped 33495  overruns 0  frame 0
        TX packets 20787361  bytes 16503009179 (16.5 GB)
        TX errors 0  dropped 0 overruns 0  carrier 0  collisions 0

> lo: flags=73<UP,LOOPBACK,RUNNING>  mtu 65536
        inet 127.0.0.1  netmask 255.0.0.0
        inet6 ::1  prefixlen 128  scopeid 0x10<host>
        loop  txqueuelen 1000  (Local Loopback)
        RX packets 3563410  bytes 162058440175 (162.0 GB)
        RX errors 0  dropped 0  overruns 0  frame 0
        TX packets 3563410  bytes 162058440175 (162.0 GB)
        TX errors 0  dropped 0 overruns 0  carrier 0  collisions 0

On the other node:

> docker0: flags=4163<UP,BROADCAST,RUNNING,MULTICAST>  mtu 1500
        inet 172.17.0.1  netmask 255.255.0.0  broadcast 172.17.255.255
        inet6 fe80::42:81ff:fe6e:ac52  prefixlen 64  scopeid 0x20<link>
        ether 02:42:81:6e:ac:52  txqueuelen 0  (Ethernet)
        RX packets 1237560  bytes 77929847 (77.9 MB)
        RX errors 0  dropped 0  overruns 0  frame 0
        TX packets 1778655  bytes 23362559991 (23.3 GB)
        TX errors 0  dropped 0 overruns 0  carrier 0  collisions 0

> enx00e04c36024f: flags=4163<UP,BROADCAST,RUNNING,MULTICAST>  mtu 1500
        inet 10.166.15.62  netmask 255.255.255.0  broadcast 10.166.15.255
        inet6 fe80::2e0:4cff:fe36:24f  prefixlen 64  scopeid 0x20<link>
        ether 00:e0:4c:36:02:4f  txqueuelen 1000  (Ethernet)
        RX packets 45537323  bytes 30931710048 (30.9 GB)
        RX errors 0  dropped 1183  overruns 0  frame 0
        TX packets 44178975  bytes 24127377081 (24.1 GB)
        TX errors 0  dropped 0 overruns 0  carrier 0  collisions 0

> lo: flags=73<UP,LOOPBACK,RUNNING>  mtu 65536
        inet 127.0.0.1  netmask 255.0.0.0
        inet6 ::1  prefixlen 128  scopeid 0x10<host>
        loop  txqueuelen 1000  (Local Loopback)
        RX packets 1397846  bytes 8924424038 (8.9 GB)
        RX errors 0  dropped 0  overruns 0  frame 0
        TX packets 1397846  bytes 8924424038 (8.9 GB)
        TX errors 0  dropped 0 overruns 0  carrier 0  collisions 0


So without the interface being renamed, the program just exits with that warning, and get stuck here https://github.com/horovod/horovod/blob/ed39bcc0cf1e0689c11c553c19cd706bf30fc8ef/horovod/runner/launch.py#L590

**Comment by zhao1157** [comment]
By the way, `ifconfig` is not installed in the docker container, is that a problem?

**Comment by chongxiaoc** [comment]
1. For common interface: if user didn't specify nics, horovod is running `get_common_interfaces` to find out a set of nics are available across all hosts. In this case, if all hosts are local, then `get_local_interfaces` would return results. I guess your execution is hanging with remote hosts, that is, https://github.com/horovod/horovod/blob/ed39bcc0cf1e0689c11c553c19cd706bf30fc8ef/horovod/runner/driver/driver_service.py#L162
, `ssh` or `KUBEFLOW_MPI_EXEC` is used to log into each host to detect nic, see: https://github.com/horovod/horovod/blob/ed39bcc0cf1e0689c11c553c19cd706bf30fc8ef/horovod/runner/util/remote.py#L29

In your case, I guess maybe `ssh` is hanging? Try `ssh` into each node?

And because you specified the nic name, so this part is skipped and you didn't see the hanging.

2. I have to correct `ifconfig` usage, it is `from psutil import net_if_addrs` which does a similar thing in python, and imported in horovod.

**Comment by zhao1157** [comment]
@chongxiaoc sshing to each other is no problem.

**Comment by chongxiaoc** [comment]
> @chongxiaoc sshing to each other is no problem.

okay, then what is `pdb` telling the hanging place? I'm interesting to see which part causes the hanging?

**Comment by zhao1157** [comment]
@chongxiaoc Here is the pdb result, which might not be useful

> > /usr/local/lib/python3.6/dist-packages/horovod/runner/launch.py(548)_run_static()
-> nics = driver_service.get_common_interfaces(settings, all_host_names,
(Pdb) next
> /usr/local/lib/python3.6/dist-packages/horovod/runner/launch.py(549)_run_static()
-> remote_host_names, fn_cache)
(Pdb) next
/usr/lib/python3.6/multiprocessing/semaphore_tracker.py:143: UserWarning: semaphore_tracker: There appear to be 5 leaked semaphores to clean up at shutdown
  len(cache))

**Comment by chongxiaoc** [comment]
Thanks for reply. On my side, I haven't done any horovod tests with docker yet, I was only focusing on tests on clusters. 
The only difference you did is to rename a nic on host1, to make it same as another nic on host2, right? for example, `enx00e04c36024f` here.

@tgaddair Do you know anything additional to check?

**Comment by zhao1157** [comment]
@chongxiaoc That's right.

**Comment by zhao1157** [comment]
@chongxiaoc Here is some more pdb output

> --Return--
> /usr/local/lib/python3.6/dist-packages/cloudpickle/cloudpickle_fast.py(563)dump()->None
-> return Pickler.dump(self, obj)
(Pdb)
> /usr/local/lib/python3.6/dist-packages/cloudpickle/cloudpickle_fast.py(103)dumps()
-> return file.getvalue()
(Pdb)
--Return--
> /usr/local/lib/python3.6/dist-packages/cloudpickle/cloudpickle_fast.py(103)dumps()->b'\x80\x04\x9...5.62:1\x94ub.'
-> return file.getvalue()
(Pdb)
> /usr/local/lib/python3.6/dist-packages/horovod/runner/common/util/codec.py(27)dumps_base64()
-> encoded = base64.b64encode(serialized)
(Pdb)
> /usr/local/lib/python3.6/dist-packages/horovod/runner/common/util/codec.py(28)dumps_base64()
-> return encoded.decode('ascii') if to_ascii else encoded
(Pdb)
--Return--
> /usr/local/lib/python3.6/dist-packages/horovod/runner/common/util/codec.py(28)dumps_base64()->'gASVOgIAAAAA...uNjI6MZR1Yi4='
-> return encoded.decode('ascii') if to_ascii else encoded
(Pdb)
> /usr/local/lib/python3.6/dist-packages/horovod/runner/driver/driver_service.py(104)_launch_task_servers()
-> if host_name not in local_host_names:
(Pdb)
> /usr/local/lib/python3.6/dist-packages/horovod/runner/driver/driver_service.py(105)_launch_task_servers()
-> command = get_remote_command(command,
(Pdb)
> /usr/local/lib/python3.6/dist-packages/horovod/runner/driver/driver_service.py(106)_launch_task_servers()
-> host=host_name,
(Pdb)
> /usr/local/lib/python3.6/dist-packages/horovod/runner/driver/driver_service.py(107)_launch_task_servers()
-> port=settings.ssh_port,
(Pdb)
> /usr/local/lib/python3.6/dist-packages/horovod/runner/driver/driver_service.py(108)_launch_task_servers()
-> identity_file=settings.ssh_identity_file)
(Pdb)
> /usr/local/lib/python3.6/dist-packages/horovod/runner/driver/driver_service.py(110)_launch_task_servers()
-> if settings.verbose >= 2:
(Pdb)
> /usr/local/lib/python3.6/dist-packages/horovod/runner/driver/driver_service.py(112)_launch_task_servers()
-> args_list.append([command])
(Pdb)
> /usr/local/lib/python3.6/dist-packages/horovod/runner/driver/driver_service.py(94)_launch_task_servers()
-> for index in range(num_hosts):
(Pdb)
> /usr/local/lib/python3.6/dist-packages/horovod/runner/driver/driver_service.py(118)_launch_task_servers()
-> threads.execute_function_multithreaded(_exec_command,
(Pdb)
> /usr/local/lib/python3.6/dist-packages/horovod/runner/driver/driver_service.py(119)_launch_task_servers()
-> args_list,
(Pdb)
> /usr/local/lib/python3.6/dist-packages/horovod/runner/driver/driver_service.py(120)_launch_task_servers()
-> block_until_all_done=False)
(Pdb)
--Return--
> /usr/local/lib/python3.6/dist-packages/horovod/runner/driver/driver_service.py(120)_launch_task_servers()->None
-> block_until_all_done=False)
(Pdb)
> /usr/local/lib/python3.6/dist-packages/horovod/runner/driver/driver_service.py(150)_driver_fn()
-> if settings.verbose >= 2:
(Pdb)
> /usr/local/lib/python3.6/dist-packages/horovod/runner/driver/driver_service.py(152)_driver_fn()
-> try:
(Pdb)
> /usr/local/lib/python3.6/dist-packages/horovod/runner/driver/driver_service.py(154)_driver_fn()
-> if settings.verbose >= 2:
(Pdb)

it exits without that warning

**Comment by zhao1157** [comment]
@chongxiaoc @tgaddair The following lines says the network interface names should be the same across nodes
https://github.com/horovod/horovod/blob/ed39bcc0cf1e0689c11c553c19cd706bf30fc8ef/horovod/runner/common/util/network.py#L161-L173

**Comment by chongxiaoc** [comment]
> @chongxiaoc @tgaddair The following lines say the network interface names should be the same across nodes
> https://github.com/horovod/horovod/blob/ed39bcc0cf1e0689c11c553c19cd706bf30fc8ef/horovod/runner/common/util/network.py#L161-L173

Yes, every host returns a set of available nics. Horovod would take a intersection of all the sets to get common ones across all nodes.

However, if user specifies --network-interfaces, and remote hosts are used, the above detection is skipped. 
See: https://github.com/horovod/horovod/blob/ed39bcc0cf1e0689c11c553c19cd706bf30fc8ef/horovod/runner/driver/driver_service.py#L240
That's the reason today `horovodrun --help` tells user to specify common nics only.

Hovorod just took the nics from command line input and uses it, and I think you already renamed two nics in docker to be same name of `enx00e04c36024f`, that is why it is working.

**Comment by zhao1157** [comment]
@chongxiaoc Is there a reason why common nics should be used in the communication? Can you point me to something related to read?

**Comment by zhao1157** [comment]
@chongxiaoc @tgaddair Without renaming the nics to be the same, after specifying `--network-interface=<nics_1>,<nics_2>`, I got the following error messages

> --------------------------------------------------------------------------
> WARNING: An invalid value was given for btl_tcp_if_include.  This
> value will be ignored.
> 
>   Local host: ubuntu
>   Value:      enx00e04c363afe
>   Message:    Unknown interface name
> --------------------------------------------------------------------------
> [1,1]<stderr>:Error in MPI_Isend(139795012727964, 1, 0x7f24a80d7e80, 0, -27, 139795415031360) (-12)
> [ubuntu:00123] *** An error occurred in MPI_Comm_dup
> [ubuntu:00123] *** reported by process [3747151873,1]
> [ubuntu:00123] *** on communicator MPI_COMM_WORLD
> [ubuntu:00123] *** MPI_ERR_INTERN: internal error
> [ubuntu:00123] *** MPI_ERRORS_ARE_FATAL (processes in this communicator will now abort,
> [ubuntu:00123] ***    and potentially your MPI job)
> [ubuntuMLPerf:00244] 1 more process has sent help message help-mpi-btl-tcp.txt / invalid if_inexclude
> [ubuntuMLPerf:00244] Set MCA parameter "orte_base_help_aggregate" to 0 to see all help / error messages

I just don't know why the nics should have the same name.

**Comment by zhao1157** [comment]
@chongxiaoc @tgaddair Does it apply to non-docker case (bare metal)? i.e. network interfaces used in the communication can have different names when running on bare metal?

**Comment by chongxiaoc** [comment]
> @chongxiaoc Is there a reason why common nics should be used in the communication? Can you point me to something related to read?

The communication backend of horovod is mpi, nccl and gloo. Common nics are passed from frontend into backend implementation-specific arguments:
- In mpi: it is `mpirun -mca btl_tcp_if_include nics`:
https://github.com/horovod/horovod/blob/3565d5fbe4c33c583892145790b7283474661dfb/horovod/runner/mpi_run.py#L160
- In nccl, it is `NCCL_SOCKET_IFNAME`:
https://github.com/horovod/horovod/blob/3565d5fbe4c33c583892145790b7283474661dfb/horovod/runner/mpi_run.py#L162

- In gloo, it is `GLOO_IFACE`:https://github.com/horovod/horovod/blob/3565d5fbe4c33c583892145790b7283474661dfb/horovod/runner/gloo_run.py#L193 

You see the difference is that only gloo uses a single nic, that is because gloo doesn't support multiple nics today, which in turn, it requires frontend to pass nics available across all hosts.

I have some in-progress work to resolve this, try to create a {host:nic_name} dict to decouple the detection of nics on hosts  (avoid using intersection of all nics), however, it was blocked by some elastic failures, because `gloo` is essential for elastic horovod, and I don't have an efficient idea to resolve it yet.

See draft: https://github.com/horovod/horovod/pull/2482

**Comment by chongxiaoc** [comment]
> @chongxiaoc @tgaddair Without renaming the nics to be the same, after specifying `--network-interface=<nics_1>,<nics_2>`, I got the following error messages
> 
> > WARNING: An invalid value was given for btl_tcp_if_include.  This
> > value will be ignored.
> > ## Local host: ubuntu
> > Value:      enx00e04c363afe
> > Message:    Unknown interface name
> > [1,1]:Error in MPI_Isend(139795012727964, 1, 0x7f24a80d7e80, 0, -27, 139795415031360) (-12)
> > [ubuntu:00123] *** An error occurred in MPI_Comm_dup
> > [ubuntu:00123] *** reported by process [3747151873,1]
> > [ubuntu:00123] *** on communicator MPI_COMM_WORLD
> > [ubuntu:00123] *** MPI_ERR_INTERN: internal error
> > [ubuntu:00123] *** MPI_ERRORS_ARE_FATAL (processes in this communicator will now abort,
> > [ubuntu:00123] ***    and potentially your MPI job)
> > [ubuntuMLPerf:00244] 1 more process has sent help message help-mpi-btl-tcp.txt / invalid if_inexclude
> > [ubuntuMLPerf:00244] Set MCA parameter "orte_base_help_aggregate" to 0 to see all help / error messages
> 
> I just don't know why the nics should have the same name.

mpirun (open mpi) throws this error, it thinks the nics are are invalid.
See open mpi implementation here:
https://github.com/open-mpi/ompi/blob/aadaf0c3c001d3dfed386931142a2d0167796ef3/opal/mca/btl/tcp/btl_tcp_component.c#L647

Maybe run with open mpi only to check the nic setup, for example:
`mpirun -np 2 -mca btl_tcp_if_include "nics_1,nics_2" hostname`

For example, here is what I got on my macbook, even "lo1" is invalid:
```
chongxiaoc@chongxiaoc-C02DL7WPMD6R horovod % mpirun -np 2 -mca btl_tcp_if_include "lo,lo1" hostname
chongxiaoc-C02DL7WPMD6R
chongxiaoc-C02DL7WPMD6R
```

**Comment by chongxiaoc** [comment]
> @chongxiaoc @tgaddair Does it apply to non-docker case (bare metal)? i.e. network interfaces used in the communication can have different names when running on bare metal?

In this case, as long as `gloo` is not used, you can specify all nics name with `--network-interface`, as my above comment, open mpi would take those as `-mca btl_tcp_if_include`, and different names across hosts are supported in this parameter in open mpi. Horovod just works as a frontend parser.

**Comment by zhao1157** [comment]
@chongxiaoc Running in docker requires the nics with the same name, while on bare metal does not?

**Comment by chongxiaoc** [comment]
> @chongxiaoc Running in docker requires the nics with the same name, while on bare metal does not?

They are same. For either case, if `--network-interface` is not used, the horovod would detect common nics internally, and if horovod cannot find due to no same name, the below exception would be thrown out:

https://github.com/horovod/horovod/blob/3565d5fbe4c33c583892145790b7283474661dfb/horovod/runner/driver/driver_service.py#L154

**Comment by zhao1157** [comment]
@okay, so in both docker and bare-metal case, the nics should have the same name. Please confirm.

**Comment by chongxiaoc** [comment]
> @chongxiaoc
> okay, so in both docker and bare-metal case, the nics should have the same name. Please confirm.

For `gloo`: yes
Other cases: don't have to: `mpirun -mca btl_tcp_if_include` and `NCCL_SOCKET_IFNAME` can filter invalid names in MPI and NCCL.

This is what I don't quite get from your feedback, since you are not using `gloo`, but seeing `mpi`  and `ssh` is hanging in nic detection. As I said above, I don't have a docker setup on my side to check your test at this moment.

Two options you could have as work around:
- Keep using renaming nics and `--network-interface` as it was working for you.
- Try master branch and install from source code to see if this issue is still there. There are some changes for nic detection code after `0.20.0`.

**Comment by zhao1157** [comment]
@chongxiaoc Thanks a lot. You are very helpful!
